File: python/30_adjlist.py

"""
Adjacency List representation of a graph in Python.
Supports adding edges and printing the adjacency list.
"""

from typing import Optional, List


class AdjNode:
    """
    Node in the adjacency list.
    """

    def __init__(self, value: int):
        self.vertex: int = value
        self.next: Optional['AdjNode'] = None


class Graph:
    """
    Undirected graph using adjacency list.
    """

    def __init__(self, num: int):
        self.V: int = num
        self.graph: List[Optional[AdjNode]] = [None] * self.V

    def add_edge(self, s: int, d: int) -> None:
        """
        Add an undirected edge between s and d.
        """
        node = AdjNode(d)
        node.next = self.graph[s]
        self.graph[s] = node

        node = AdjNode(s)
        node.next = self.graph[d]
        self.graph[d] = node

    def print_graph(self) -> None:
        """
        Print the adjacency list of the graph.
        """
        for i in range(self.V):
            print(f"Vertex {i}:", end="")
            temp = self.graph[i]
            while temp:
                print(f" -> {temp.vertex}", end="")
                temp = temp.next
            print()


if __name__ == "__main__":
    V = 5
    graph = Graph(V)
    graph.add_edge(0, 1)
    graph.add_edge(0, 2)
    graph.add_edge(0, 3)
    graph.add_edge(1, 2)
    graph.print_graph()

File: java/30_adjlist.java

/**
 * Adjacency-list representation (undirected) demo.
 *
 * This file demonstrates building a simple undirected graph using an
 * adjacency list backed by {@link java.util.ArrayList}. The example output
 * and algorithmic behaviour are preserved; the code is refactored for
 * readability and uses the {@link java.util.List} interface.
 */

import java.util.List;
import java.util.ArrayList;

class Graph {

    /**
     * Add an undirected edge between {@code s} and {@code d} in the
     * adjacency list {@code am}.
     *
     * @throws IndexOutOfBoundsException if s or d are out of range for am
     */
    static void addEdge(List<List<Integer>> am, int s, int d) {
        am.get(s).add(d);
        am.get(d).add(s);
    }

    public static void main(String[] args) {

        // Create the graph
        int V = 5;
        List<List<Integer>> am = new ArrayList<>(V);

        for (int i = 0; i < V; i++) {
            am.add(new ArrayList<Integer>());
        }

        // Add edges
        addEdge(am, 0, 1);
        addEdge(am, 0, 2);
        addEdge(am, 0, 3);
        addEdge(am, 1, 2);

        printGraph(am);
    }

    // Print the graph (keeps the original printed layout)
    static void printGraph(List<List<Integer>> am) {
        for (int i = 0; i < am.size(); i++) {
            System.out.println("\nVertex " + i + ":");
            for (Integer neighbor : am.get(i)) {
                System.out.print(" -> " + neighbor);
            }
            System.out.println();
        }
    }
}